compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -793,6 +793,10 @@ fn convert_enum_variant_types(tcx: TyCtxt<'_>, def_id: DefId) {
 fn find_field(tcx: TyCtxt<'_>, (def_id, ident): (DefId, Ident)) -> Option<FieldIdx> {
     tcx.adt_def(def_id).non_enum_variant().fields.iter_enumerated().find_map(|(idx, field)| {
         if field.is_unnamed() {
+            // Users can't write `expr._`, but derive macros can
```

Comment:
The good thing is, only *built-in* (HIR-based) macros have this capability. Neither decl macros nor proc macros can construct an ``hir::ExprKind::Field(…, `_`)`` since they're based on token streams and thus parser would bail out when encountering `$expr._`.

Reply:
So I guess it could make sense to add a `debug_assert_ne!` inside `rustc_expand::base::ExtCtxt::xpr_field`?